an-introduction-to-unit-testing.md

```diff
@@ -0,0 +1,852 @@
+
+### Unit tests vs integration tests
+
+- Can I run and get results from the test I wrote two weeks, months, or years ago?
+- Can any member of my team run and get results from tests I wrote two months ago?
+- Can I run all the tests I’ve written in no more than a few minutes?
+- Can I run all the tests I’ve written at the push of a button?
+- Can I write a basic test in no more than a few minutes?
+
+If the answer to one or more of the questions above is no, then I have stepped into the realm of integration testing.
+
+### Unit test definition 3
+
+A unit test is an automated piece of code that invokes the unit of work being tested, and then checks some assumptions about a single end result of that unit. A unit test is almost always written using a unit testing framework. It can be written easily and runs quickly. It’s trustworthy, readable, and maintainable. It’s consistent in its results as long as production code hasn’t been changed.
+
+### Test-driven development (TDD)
+
+TDD is a practice that involves writing unit tests before the production code is written, writing a failing test to prove that code or functionalities are missing, and then making the test pass by writing production code that meets the expectations.
+
+When refactoring your code (refactoring means changing a piece of code without changing its functionality), seeing a test fail, and then seeing it pass without changing the test, you’re basically testing the test itself too.
+
+### Types of unit testing
+
+**Value-based testing** checks the value returned from a function.
+
+**State-based testing** is about checking for noticeable behavior changes.
+
+**Interaction testing** is testing how an object sends messages (i.e. calls methods) to other objects. You use interaction testing when calling another object is the end result of a specific unit of work.
+
+## Unit testing in practice
+
+This section introduces the xUnit testing frameworks.
+
+### Unit testing frameworks
+
+Unit tests are written as code, using libraries from the unit testing framework. The tests are run from a separate unit testing tool or inside the IDE, and the results are reviewed (either as output text, the IDE, or the unit testing framework application UI) by the developer or an automated build process.
+
+### The xUnit frameworks
+
+xUnit is a collective name for several unit testing frameworks that derive their structure and functionality from Smalltalk’s SUnit, designed by Kent Beck in 1998.
+
+### Naming convention
+
+The name of a unit test usually includes the name of the class, the name of the unit of work, and the expected behavior.
+
+Example:  `OpticalSensor_IsRedBloodCell_ReturnsFalse()`
```

Comment:
There are many naming conventions, some projects for example are using a "given..._when..._then..." format